File: Sort/code/php/Sort.php
<?php
/**
 * 排序算法实现
 * @author phachon@163.com
 */

class Sort {

	/**
	 * Sort constructor.
	 */
	public function __construct() {}

	/**
	 * 选择排序
	 * @param array $data
	 * @return array
	 */
	public function select($data = array()) {
		$count = count($data);
		for ($i = 0; $i < $count; $i++) {
			// 每次都假设第一个位置为最小的元素
			$min = $i;
			for ($j = $i + 1; $j < $count; $j ++) {
				if ($data[$j] < $data[$min]) {
					$min = $j;
				}
			}
			// 判断最小的数是否还是 $i, 否则交换位置
			if ($min != $i) {
				$tmp = $data[$i];
				$data[$i] = $data[$min];
				$data[$min] = $tmp;
 			}
		}
		return $data;
	}

	/**
	 * 冒泡排序
	 * @param array $data
	 * @return array
	 */
	public function bubbling($data = array()) {
		$count = count($data);
		for ($i = 0; $i < $count; $i++) {
			for($j = 0; $j < $count - $i - 1; $j++) {
				if ($data[$j] > $data[$j + 1]) {
					$tmp = $data[$j + 1];
					$data[$j + 1] = $data[$j];
					$data[$j] = $tmp;
				}
			}
		}
		return $data;
	}

	/**
	 * 直接插入排序
	 * @param array $data
	 * @return array
	 */
	public function insert($data = array()) {
		$count = count($data);
		for ($i = 1; $i < $count; $i++) {
			$insert = $data[$i];
			for ($j = $i - 1; $j >= 0; $j--) {
				if ($insert < $data[$j]) {
					$data[$j+1] = $data[$j];
				}else {
					break;
				}
			}
			$data[$j+1] = $insert;
		}
		return $data;
	}

	/**
	 * 二分插入排序
	 * @param array $data
	 * @return array
	 */
	public function binaryInsert($data = array()) {
		$count = count($data);
		for ($i = 1; $i < $count; $i++) {
			$insert = $data[$i];
			if ($insert < $data[$i-1]) {
				// 二分查找插入
				// 0 ~ (i -1)
				$left = 0;
				$right = $i - 1;
				while ($left <= $right) {
					$middle = intval(($left + $right) / 2);
					if ($insert < $data[$middle]) {
						$right = $middle - 1;
					} else {
						$left = $middle + 1;
					}
				}
				// 找到位置，位置之后元素后移
				for ($k = $i - 1; $k >= $left; $k--) {
					$data[$k + 1] = $data[$k];
				}
				$data[$left] = $insert;
			}
		}


		return $data;
	}

	/**
	 * 快速排序
	 * @param array $data
	 * @return array
	 */
	public function quick($data = array()) {
		$count = count($data);
		if(count($data) <= 1) {
			return $data;
		}
		// 基准数
		$key = $data[0];
		$bigList = [];
		$smallList = [];

		// 以基准数为界限分隔
		for ($i = 1; $i < $count; $i++) {
			if ($data[$i] > $key) {
				array_push($bigList, $data[$i]);
			}else {
				array_push($smallList, $data[$i]);
			}
		}

		$bigList = $this->quick($bigList);
		$smallList = $this->quick($smallList);

		return array_merge($smallList, array($key), $bigList);

	}

	/**
	 * 归并排序
	 * @param array $data
	 * @return array
	 */
	public function merge($data = array()) {
		$start = 0;
		$end = count($data) - 1;
		$this->_mergeSort($start, $end, $data);
		return $data;
	}

	/**
	 * 递归拆分并合并
	 * @param array $data
	 * @param $start
	 * @param $end
	 */
	private function _mergeSort($start, $end, array &$data) {
		if ($start < $end) {
			$middle = floor(($start + $end) / 2);

			// 左半部分
			$this->_mergeSort($start, $middle, $data);
			// 右半部分
			$this->_mergeSort($middle + 1, $end, $data);

			// 归并操作
			$this->_merge($start, $middle, $end, $data);
		}
	}

	/**
	 * 将一个数组中的两个相邻有序区间合并成一个
	 * @param $start
	 * @param $middle
	 * @param $end
	 * @param $data
	 */
	private function _merge($start, $middle, $end, array &$data) {
		$i = $start;
		$j = $middle + 1;
		$k = $start;
		$tmpArr = [];

		while($i != $middle + 1 && $j != $end+1) {
			if($data[$i] < $data[$j]) {
				$tmpArr[$k++] = $data[$i++];
			}else {
				$tmpArr[$k++] = $data[$j++];
			}
		}

		// 第一个序列的剩余部分添加到数组
		while ($i != $middle+1) {
			$tmpArr[$k++] = $data[$i++];
		}
		// 第二个序列的剩余部分添加到数组
		while ($j != $end+1) {
			$tmpArr[$k++] = $data[$j++];
		}

		// 改变 data
		for($i = $start; $i <= $end; $i++) {
			$data[$i] = $tmpArr[$i];
		}
	}

	/**
	 * 希尔排序
	 * @param $data
	 * @return mixed
	 */
	public function shell($data) {
		$count = count($data);
		// 分组的大小
		$number = 2;

		// gap为步长，每次减为原来的一半。
		for ($gap = floor($count/$number); $gap > 0; $gap = floor($gap/$number)) {
			// 对 gap 个分组数组执行直接插入排序
			for ($i = 0; $i < $gap; $i++) {
				// 直接插入排序
				for ($j = $i; $j < $count; $j += $gap) {
					$insert = $data[$j];
					for ($z = $j - $gap; $z >= $i; $z -= $gap) {
						if ($insert < $data[$z]) {
							// 移动元素到后面
							$data[$z+$gap] = $data[$z];
						}else {
							break;
						}
					}
					$data[$z+$gap] = $insert;
				}
			}
		}

		return $data;
	}

	/**
	 * @todo 堆排序
     * 第一个非叶子结点 count(arr)/2-1
	 * @param array $data
	 */
	public function heap($data = array()) {

		// 1. 构建大顶堆
		for ($i = floor(count($data) / 2) - 1; $i >= 0; $i--) {
			// 从下至上，从右至左调整
			$this->_makeMaxHeap($data);
		}

		// 2. 交换堆顶元素和末尾元素，重新调整堆结构
	}

	/**
	 * 调整大顶堆
	 * 大顶堆：arr[i] >= arr[2i+1] && arr[i] >= arr[2i+2]
	 * @param array $data
	 * @param $i
	 */
	private function _makeMaxHeap($i, $data = array()) {
		$current = $data[$i];

		for ($k = $i*2 + 1; $k < count($data); $k = $i*2 + 1) { //从i结点的左子结点开始，也就是2i+1处开始
			if($k+1 < count($data) && $data[$k] < $data[$k+1]) { //如果左子结点小于右子结点，k 指向右子结点
				$k++;
			}

			if($data[$k] > $current){//如果子节点大于父节点，将子节点值赋给父节点（不用进行交换）
				$data[$i] = $data[$k];
				$i = $k;
			}else{
				break;
			}
		}

		$data[$i] = $current;//将temp值放到最终的位置
	}

	/**
	 * 计数排序
	 * @param array $data
	 * @return array
	 */
	public function count($data = array()) {

		$count = count($data);
		// 输出数组
		$outArr = [];
		// 辅助数组
		$tmpArr = [];
		// 查找最大的数
		for ($i = 0; $i < $count - 1; $i++) {
			if($data[$i] > $data[$i+1]) {
				$tmp = $data[$i];
				$data[$i] = $data[$i+1];
				$data[$i+1] = $tmp;
			}
		}
		$max = $data[$count-1];

		// 初始化辅助数组
		for ($i = 0; $i < $max+1; $i++) {
			$tmpArr[$i] = 0;
		}
		// 初始化输出数组
		for ($i = 0; $i < $count; $i++) {
			$outArr[$i] = 0;
		}
		// 对次数进行统计
		for ($i = 0; $i < $count; $i++) {
			$tmpArr[$data[$i]]++;
		}

		// 累加求和
		for ($i = 1; $i < count($tmpArr); $i++) {
			$tmpArr[$i] = $tmpArr[$i] + $tmpArr[$i-1];
		}

		// 循环遍历到输出数组
		for ($i = 0; $i < $count; $i++) {
			// 记得减一，因为是从 0 开始
			$outArr[$tmpArr[$data[$i]]-1] = $data[$i];
			// 计数减1
			$tmpArr[$data[$i]] --;
		}
		return $outArr;
	}

	/**
	 * 桶排序
	 * @param $data
	 * @return array
	 */
	public function bucket($data) {
		$outArr = [];
		$tmpArr = [];
		$count = count($data);

		// 查找最大的值
		for ($i=0; $i < $count - 1; $i++) {
			if ($data[$i] > $data[$i+1]) {
				$tmp = $data[$i+1];
				$data[$i+1] = $data[$i];
				$data[$i] = $tmp;
			}
		}
		$max = $data[$count-1];

		// 初始化桶，用 11 个桶, 桶保存的是链表, 每个桶的数据范围是 max/10
		for ($i = 0; $i < 11; $i++) {
			$tmpArr[$i] = new LinkedNode(); // 空的头节点
		}

		// 开始为桶加数据
		for ($i = 0; $i < $count; $i++) {
			//查找是在哪个桶
			$n = floor($data[$i]/($max/10));
			// 查找到桶之后对桶中的数据排序并插入,最好用链表，不需要移动元素
			$b = $tmpArr[$n];
			while ($b->next != NULL) {
				if($data[$i] <= $b->next->data) {
					break;
				}else {
					$b = $b->next;
				}
			}
			$node = new LinkedNode($data[$i]);
			$node->next = $b->next;
			$b->next = $node;
		}

		// 将桶里的数据倒出来
		for ($i = 0; $i < 11; $i++) {
			$current = $tmpArr[$i]->next;
			while ($current != NULL) {
				array_push($outArr, $current->data);
				$current = $current->next;
			}
		}

		return $outArr;
	}

	/**
	 * 基数排序
	 * @param $data
	 * @return array
	 */
	public function radix($data) {

		$count = count($data);
		// 查找数组的最大值
		for ($i = 0; $i < $count-1; $i ++) {
			if ($data[$i] > $data[$i+1]) {
				$tmp = $data[$i+1];
				$data[$i+1] = $data[$i];
				$data[$i] = $tmp;
			}
		}
		$max = $data[$count-1];
		//获取最大位数
		$len = strlen((string)$max);

		return $this->_loopRadix($data, 1, $len);
	}

	/**
	 * 循环
	 * @param $arr
	 * @param $digit
	 * @param $len
	 * @return array
	 */
	private function _loopRadix(array $arr, $digit, $len) {
		if ($digit >= $len) {
			return $arr;
		}

		// 利用计数排序对该位数的数字排序

		// 构造临时桶数组
		$tmpArr = [];
		for ($i=0; $i < 10; $i++) {
			$tmpArr[$i] = 0;
		}
		$outArr = [];
		for ($i = 0; $i < count($arr); $i++) {
			$outArr[$i] = 0;
		}

		// 桶放数据, 出现的次数加1
		for ($i = 0; $i < count($arr); $i++) {
			$tmpArr[$this->_getDigit($arr[$i], $digit)]++;
		}

		// 转化成下标索引
		for ($i = 1; $i < 10; $i++) {
			$tmpArr[$i] = $tmpArr[$i] + $tmpArr[$i-1];
		}
		// 输出到新数组, 注意：必须倒序输出！
		for ($i = count($arr) -1; $i >= 0; $i--) {
			$d = $this->_getDigit($arr[$i], $digit);
			$outArr[$tmpArr[$d] - 1] = $arr[$i];
			$tmpArr[$d]--;
		}

		return $this->_loopRadix($outArr, $digit + 1, $len);
	}

	/**
	 * 获取某一个数字第 n 位的数字
	 * @param $number 数字
	 * @param $n 第几位 从右向左数
	 * @return int
	 */
	private function _getDigit($number, $n) {
		$pow = pow(10, $n - 1);
		return ($number/$pow) % 10;
	}

	public function printArray($data) {
		echo implode(",", $data)."\r\n";
	}
}

File: Sort/code/golang/sort.go
package main

import (
	"math"
	"fmt"
)

// 排序算法的实现

type Sort struct {

}

// 选择排序
func (s *Sort) Select(data1 []int) ([]int) {
	data := data1
	count := len(data)
	for i := 0; i < count; i++ {
		min := i
		for y := i + 1; y < count; y++ {
			if data[y] < data[min] {
				min = y
			}
		}

		if min != i {
			tmp := data[min]
			data[min] = data[i]
			data[i] = tmp
		}
	}
	return  data
}

// 冒泡排序
func (s *Sort) Bubbling(data []int) []int {
	count := len(data)

	for i := 0; i < count; i++ {
		for j := 0; j < count - i -1; j++ {
			if data[j] > data[j+1] {
				tmp := data[j]
				data[j] = data[j+1]
				data[j+1] = tmp
			}
		}
	}
	return data
}

// 直接插入排序
func (s *Sort) Insert(data []int) []int {
	count := len(data)

	for i := 1; i < count; i++ {
		insert := data[i]
		var j int
		for j = i - 1; j >= 0; j-- {
			if insert < data[j]  {
				data[j+1] = data[j]
			}else {
				break
			}
		}
		data[j+1] = insert
	}
	return data
}

// 二分插入排序
func (s *Sort) BinaryInsert(data []int) []int {
	count := len(data)

	for i := 1; i < count; i++ {
		left := 0
		right := i - 1
		insert := data[i]

		for left <= right {
			middle := (left + right) / 2
			if insert > data[middle] {
				left = middle + 1
			}else {
				right = middle - 1
			}
		}

		for j := i-1; j >= left; j-- {
			data[j+1] = data[j]
		}
		data[left] = insert
	}
	return  data
}

// 快速排序
func (s *Sort) Quick(data []int) []int {
	count := len(data)
	if count <= 1 {
		return data
	}
	bigList := []int{}
	smallList := []int{}
	// 基准数
	tmp := data[0]

	for i := 1; i < count; i++ {
		if data[i] >= tmp {
			bigList = append(bigList, data[i])
		}else {
			smallList = append(smallList, data[i])
		}
	}

	bigList = s.Quick(bigList)
	smallList = s.Quick(smallList)

	res := append(smallList, tmp)
	res = append(res, bigList...)
	return res
}

// 归并排序
// 分而治之
func (s *Sort) Merge(data []int) []int {
	s.merge(0, len(data) - 1, data)
	return data
}

func (s *Sort) merge(start int, end int, data []int) {
	if start < end {
		middle := (start + end) / 2
		s.merge(start, middle, data)
		s.merge(middle+1, end, data)
		s.mergeList(start, middle, end, data)
	}
}

// 合并两个联系的数组
func (s *Sort) mergeList(start int, middle int, end int, data []int) {
	tmp := make([]int, len(data))
	lStart := start
	lEnd := middle

	rStart := middle + 1
	rEnd := end

	k := 0
	for (lStart <= lEnd) && (rStart <= rEnd) {
		if data[lStart] > data[rStart] {
			tmp[k] = data[rStart]
			rStart++
			k++
		}else {
			tmp[k] = data[lStart]
			lStart++
			k++
		}
	}
	// 检查是否剩余
	for lStart <= lEnd {
		tmp[k] = data[lStart]
		lStart++
		k++
	}
	for rStart <= rEnd {
		tmp[k] = data[rStart]
		rStart++
		k++
	}
	//移动元素
	for x := 0; x < k; x++ {
		data[start + x] = tmp[x]
	}
}

// 希尔排序
func (s *Sort) Shell(data []int) []int {

	count := len(data)
	// 第一次分组数
	size := 2

	// gap 为步长，每次减为原来的一半
	for gap := count / size ; gap > 0; gap /= size {
		// 对每组进行直接插入排序后再分
		for i := 0; i < gap; i++ {
			for x := i; x < count; x += gap {
				insert := data[x]
				var y int
				for y = x - gap; y >= i; y -= gap {
					if insert < data[y] {
						data[y+gap] = data[y]
					}else {
						break
					}
				}
				data[y+gap] = insert
			}
		}
	}
	return  data
}

// 计数排序
func (s *Sort) Count(data []int) []int {
	count := len(data)
	// 计算最大的数
	for i := 0; i < count - 1; i++ {
		if data[i] > data[i+1] {
			tmp := data[i+1]
			data[i+1] = data[i]
			data[i] = tmp
		}
	}
	max := data[count - 1]
	// 初始化桶
	bucket := make([]int, max+1)
	// 出现次数放入桶
	for x := 0; x < count; x++ {
		bucket[data[x]]++
	}
	// 对桶的数据累加
	for x := 1; x < max + 1; x++ {
		bucket[x] = bucket[x] + bucket[x-1]
	}
	out := make([]int, count)
	// 输出到新数组
	for x := 0; x < count; x++ {
		out[bucket[data[x]]-1] = data[x]
		bucket[data[x]]--
	}
	return out
}

// 单向链表节点
type LinkedNode struct {
	Next *LinkedNode
	Data int
}

func newLinkedNode() *LinkedNode {
	return &LinkedNode{
		Data: 0,
	}
}

// 桶排序
func (s *Sort) Bucket(data []int) []int {
	count := len(data)
	// 计算最大的数
	for i := 0; i < count - 1; i++ {
		if data[i] > data[i+1] {
			tmp := data[i+1]
			data[i+1] = data[i]
			data[i] = tmp
		}
	}
	max := data[count - 1]
	// 桶的数量
	size := 10
	// 初始化桶，桶中是单向链表，初始化头结点
	bucket := make([]*LinkedNode, size+1)
	for i := 0; i < len(bucket); i++ {
		bucket[i] = newLinkedNode()
	}

	// 满足条件的数据放入桶中
	for i := 0; i < count; i++ {
		bList := bucket[data[i] / (max /size)]
		// 对桶中的数排序
		current := bList
		for current.Next != nil {
			if data[i] < current.Next.Data {
				break
			}else {
				current = current.Next
			}
		}
		// 移动链表节点
		node := &LinkedNode{
			Next: current.Next,
			Data: data[i],
		}
		current.Next = node
	}
	out := []int{}
	// 输出数据都新数组
	for x := 0; x < len(bucket); x++ {
		bList := bucket[x]
		current := bList.Next
		for current != nil {
			out = append(out, current.Data)
			current = current.Next
		}
	}
	return out
}

// 基数排序
// 按位数分次排序
func (s *Sort) Radix(data []int) []int {
	count := len(data)
	//查找最大值
	for i := 0; i < count - 1; i++ {
		if data[i] > data[i+1] {
			tmp := data[i+1]
			data[i+1] = data[i]
			data[i] = tmp
		}
	}
	max := data[count - 1]
	// 最多的位数
	maxDigit := len(fmt.Sprintf("%d", max))

	// 每一位循环处理
	for digit := 1; digit <= maxDigit; digit++ {
		tmp := make([]int, max+1)
		for x := 0; x < count; x++ {
			d := s.digit(data[x], digit)
			tmp[d]++
		}
		// 对位数进行计数排序
		for y := 1; y < len(tmp); y++ {
			tmp[y] = tmp[y] + tmp[y - 1]
		}
		// 输出，必须倒着输出
		out := make([]int, count)
		for z := count-1; z >= 0; z-- {
			d := s.digit(data[z], digit)
			out[tmp[d] - 1] = data[z]
			tmp[d]--
		}
		data = out
	}
	return data
}

// 返回某个数第几位的数字
// e: 1346 第 1 位数字为 6
func (s *Sort) digit(number int, n int) int {
	x := math.Pow(10, float64(n - 1))
	return (number / int(x)) % 10
}

// 堆排序
func (s *Sort) Heap(data []int) []int {
	count := len(data)
	// 循环构建大顶堆
	for i := 0; i < count; i++ {
		// 构建大顶堆
		s.makeMaxHeap(count - i - 1, data)
		// 交换元素, 大顶堆的第一个元素和最后一个元素交换
		tmp := data[0]
		data[0] = data[count - i - 1]
		data[count - i - 1] = tmp
	}
	return data
}

// 构建大顶堆
func (s *Sort) makeMaxHeap(lastIndex int, data []int) {

	// 从最后一个元素的父节点开始
	for i := (lastIndex - 1) / 2; i >= 0; i-- {
		current := i
		for (current * 2 + 1) <= lastIndex {
			// 假设左节点最大
			bigIndex := (current * 2) + 1
			// 存在右节点
			if lastIndex > bigIndex {
				//比较左右节点
				if data[bigIndex+1] > data[bigIndex] {
					bigIndex++
				}
			}
			// 比较最大的叶子节点和当前节点
			if data[current] < data[bigIndex] {
				tmp := data[bigIndex]
				data[bigIndex] = data[current]
				data[current] = tmp

				current = bigIndex
			}else {
				break
			}
		}
	}
}